File: readability/readability.c
#include <cs50.h>
#include <stdio.h>
#include <ctype.h>
#include <math.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
    string text = get_string("Text: ");

    float charactercount = count_letters(text);
    int wordcount = count_words(text);
    float sentencecount = count_sentences(text);
    int offset = sentencecount - 1;
    int truewordcount = wordcount - offset;

    float L = (charactercount / truewordcount) * 100;
    float S = (sentencecount / truewordcount) * 100;

    float grade = (0.0588 * L) - (0.296 * S) - 15.8;

    int roundedgrade = round(grade);

    if (roundedgrade < 1)
    {
        printf("Before Grade 1\n");
    }

    else if (roundedgrade >= 16)
    {
        printf("Grade 16+\n");
    }

    else
    {
        printf("Grade %i\n", roundedgrade);
    }
}

int count_letters(string text)
{
    int i = 0;
    int count = 0;
    while (text[i] != '\0')
    {
        if (isalnum(text[i]))
        {
            count++;
            i++;
        }
        else
        {
            i++;
        }
    }
    return count;
}

int count_words(string text)
{
    int i = 0;
    int count = 0;
    while (text[i] != '\0')
    {
        if (text[i] == ' ' || text[i] == '.' || text[i] == '?' || text[i] == '!')
        {
            count++;
            i++;
        }
        else
        {
            i++;
        }
    }
    return count;
}

int count_sentences(string text)
{
    int i = 0;
    int count = 0;
    while (text[i] != '\0')
    {
        if (text[i] == '.' || text[i] == '?' || text[i] == '!')
        {
            count++;
            i++;
        }
        else
        {
            i++;
        }
    }
    return count;
}

File: sentimental-readability/readability.py
from cs50 import get_string
import re
import string


def main():
    text = get_string("Text: ")
    letter_count = count_letters(text)
    word_count = count_words(text)
    sentence_count = count_sentences(text)
    grade = calculate_grade(letter_count, word_count, sentence_count)
    if grade < 1:
        print("Before Grade 1")
    elif grade >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {grade}")


def count_letters(text):
    new_text = text.translate(str.maketrans('', '', string.punctuation))
    words = re.split(r'\s', new_text)
    letter_count = 0
    for word in words:
        letter_count = letter_count + len(word)
    return letter_count


def count_words(text):
    words = re.split(r'\s', text)
    word_count = len(words)
    return word_count


def count_sentences(text):
    sentences = re.split(r'\.|\?|\!', text)
    sentence_count = len(sentences)
    return sentence_count - 1


def calculate_grade(letter_count, word_count, sentence_count):
    L = (letter_count / word_count) * 100
    S = (sentence_count / word_count) * 100
    grade = 0.0588 * L - 0.296 * S - 15.8
    return round(grade)


main()